Support performing merge appends and delete files on branches

amogh-jahagirdar opened the pull request:
I realized as I was implementing some of the snapshot producing operations for branches, that one operation that could be handled in a straightforward manner was delete files. It also helps with writing some of the test cases for the implementation of branch operations because for some of those tests we want to delete files on a given branch. So separating out this PR.

cc: @rdblue @jackye1995 @namrathamyske

namrathamyske commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -300,6 +300,70 @@ public void testDeleteCaseSensitivity() {
+    Assert.assertEquals("Should have 1 manifest", 1, deleteSnapshot.allManifests(FILE_IO).size());
+    validateManifestEntries(
+        deleteSnapshot.allManifests(FILE_IO).get(0),
+        ids(initialSnapshot.snapshotId(), deleteSnapshot.snapshotId()),
```
In test cases initialSnapshot is always behind deleteSnapshot. Should there be a check when it's ahead of deleteSnapshot ? Not sure if it's necessary.

amogh-jahagirdar replied:
Sorry not sure what you mean by "ahead" and "behind"? Do you mean if we should validate the state of main after the branch writes? 

In that case, it makes sense but I don't think its particularly necessary because the existing tests in SnapshotProducer should make sure we're committing snapshots to the correct branch in the first place meaning the existing tests would guarantee us that the operations we perform are done on the correct lineage, and existing lineages would be untouched. That being said, no harm in adding more validations! @jackye1995 @rdblue any thoughts?

jackye1995 replied:
Yes agree with what Amogh says, but we can always add a test to verify the behavior.

rdblue replied:
I think these tests are specific enough that we don't need to validate the order of commits. Entries with `Status.DELETED` will only appear in a single manifest and are dropped if the manifest is rewritten. So validating DELETE entries can only succeed if the delete operation produced the manifest.

rdblue commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -300,6 +300,70 @@ public void testDeleteCaseSensitivity() {
+  }
+
+  @Test
+  public void testDeleteWithRowFilterWithCombinedPredicatesOnBranch() {
```
Similar to the PR for `RowDelta`, can we parameterize this so that we test all of the existing cases on a branch that is independent from `main`? Then we only need tests for modifications to both main and a separate branch (to validate that they evolve independently).

We'll also need to update snapshot expiration tests to ensure that file deletion happens as expected.

rdblue replied:
Here's the comment on RowDelta: https://github.com/apache/iceberg/pull/5234/files#r953201851

amogh-jahagirdar replied:
I've parameterized the tests, it does seem like for meaningful tests to be written here we also need to have an implementation for MergeAppend. So I added that.

amogh-jahagirdar commented:
The tests for delete files depend on merge appends for some test cases, and vice versa. So to have meaningful branch write tests, I've added the implementation for merge append as well and updated all the test cases to be parameterized so existing cases work for main and branch cases. I still need to follow up on updated expiration tests and then the tests which can test evolution of different branches independently.

namrathamyske commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -68,38 +68,46 @@ public class TestDeleteFiles extends TableTestBase {
   public static Object[] parameters() {
-    return new Object[] {1, 2};
+    return new Object[][] {
+      new Object[] {1, "main"},
```
I am little confused why formatVersion parameters were removed.

amogh-jahagirdar replied:
That's correct, we test for every combination of format version + branch. I'll need update the @Parameters so that we can get nicer test names. The second parameter gets injected as an argument in the test file constructor.

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/TableTestBase.java:
```diff
@@ -349,6 +349,33 @@ void validateSnapshot(Snapshot old, Snapshot snap, long sequenceNumber, DataFile
     validateSnapshot(old, snap, (Long) sequenceNumber, newFiles);
   }
 
+  Snapshot commit(Table tbl, SnapshotUpdate snapshotUpdate, String branch) {
+    Snapshot snapshot = null;
+    if (branch.equals(SnapshotRef.MAIN_BRANCH)) {
+      snapshotUpdate.commit();
+      snapshot = tbl.currentSnapshot();
+    } else {
+      ((SnapshotProducer) snapshotUpdate.toBranch(branch)).commit();
+      snapshot = tbl.snapshot(tbl.refs().get(branch).snapshotId());
+    }
+    return snapshot;
+  }
+
+  Snapshot apply(SnapshotUpdate snapshotUpdate, String branch) {
+    if (branch.equals(SnapshotRef.MAIN_BRANCH)) {
+      return ((SnapshotProducer) snapshotUpdate).apply();
+    } else {
+      return ((SnapshotProducer) snapshotUpdate.toBranch(branch)).apply();
+    }
+  }
```
I could combine these into a common method apply(Table tl, SnapshotUpdate snapshotUpdate, String branch, boolean commit) and then only commit if commit is true. but for now opting to keep it explicit and separate, unless there's any objections

rdblue commented on core/src/test/java/org/apache/iceberg/TableTestBase.java:
```diff
@@ -349,6 +349,35 @@ void validateSnapshot(Snapshot old, Snapshot snap, long sequenceNumber, DataFile
     validateSnapshot(old, snap, (Long) sequenceNumber, newFiles);
   }
 
+  Snapshot commit(Table tbl, SnapshotUpdate snapshotUpdate, String branch) {
```
Minor: we avoid unnecessary abbreviations, like `tbl` in place of `table`. It's often more awkward to type the shortened version and it makes the code harder to read. I think in this case both are true. `t` and `b` right after one another are hard to type, but I it's easy to type `tab`.

amogh-jahagirdar replied:
good point, yeah in this case it makes sense just to use the actual name

amogh-jahagirdar replied:
I forgot why I did tbl, it was because table was already used and was failing the hidden field checkstyle. I updated to ignore the checkstyle warning in favor of more readable variable names.

rdblue commented on core/src/test/java/org/apache/iceberg/TableTestBase.java:
```diff
@@ -349,6 +349,35 @@ void validateSnapshot(Snapshot old, Snapshot snap, long sequenceNumber, DataFile
+      snapshotUpdate.commit();
+      snapshot = tbl.currentSnapshot();
+    } else {
+      ((SnapshotProducer) snapshotUpdate.toBranch(branch)).commit();
```
Why not just pass a `SnapshotProducer` to this method? That would eliminate casting right?

amogh-jahagirdar replied:
MergeAppend doesn't implement SnapshotProducer

rdblue commented on core/src/test/java/org/apache/iceberg/TableTestBase.java:
```diff
@@ -349,6 +349,35 @@ void validateSnapshot(Snapshot old, Snapshot snap, long sequenceNumber, DataFile
+    }
+  }
+
+  Snapshot currentSnapshot(Table tbl, String branch) {
```
How about `latest` rather than `current`? I think that describes it better since `current` has a specific meaning (the latest on the main branch).

amogh-jahagirdar replied:
updated to `latest`

rdblue commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -68,38 +68,46 @@ public class TestDeleteFiles extends TableTestBase {
   @Test
   public void testMultipleDeletes() {
-    table.newAppend().appendFile(FILE_A).appendFile(FILE_B).appendFile(FILE_C).commit();
+    Snapshot append =
```
Style (and minor): I'd prefer to prepare the change and then pass it in to avoid this formatting. I think it would be cleaner and more readable like this:

```java
SnapshotProducer action = table.newAppend().appendFile(FILE_A).appendFile(FILE_B).appendFile(FILE_C);
commit(table, action, branch);

Snapshot append = latestSnapshot(readMetadata(), branch);
```

I also prefer to call `latestSnapshot` separately to get the `Snapshot`. We don't need to put too much on `commit`.

rdblue replied:
Thinking about this a bit more, returning the committed `Snapshot` from `commit` is fine. Up to you what to do here, I'm probably being pedantic.

amogh-jahagirdar replied:
Sure, I left it as is for now.

rdblue commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -68,38 +68,46 @@ public class TestDeleteFiles extends TableTestBase {
+            branch);
 
     Assert.assertEquals("Metadata should be at version 1", 1L, (long) version());
-    Snapshot append = readMetadata().currentSnapshot();
```
This is actually different validation behavior because the `readMetadata()` method doesn't go through the table. It goes directly to the backing store to get the metadata. I'm not sure that this is a problem, but I'd probably lean toward keeping the existing behavior. We can just make a version of `latestSnapshot` that accepts `TableMetadata` as returned by `readMetadata()`.

amogh-jahagirdar replied:
I've updated with an additional latestSnapshot method which takes in `TableMetadata` . The cases which were using it before, are now using this method instead of through the table.

rdblue commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -300,6 +322,34 @@ public void testDeleteCaseSensitivity() {
+
+    // Verify B and C on testBranch
+    validateManifestEntries(
+        testBranchTip.allManifests(FILE_IO).get(0),
```
It's much better to use something like `Iterables.getOnlyElement` rather than `.get(0)` to validate your assumptions about there being only one manifest.

rdblue commented on core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:
```diff
@@ -300,6 +322,34 @@ public void testDeleteCaseSensitivity() {
+    Snapshot initialSnapshot = table.currentSnapshot();
+    // Delete A on test branch
+    table.newDelete().deleteFile(FILE_A).toBranch(testBranch).commit();
+    Snapshot testBranchTip = table.snapshot(testBranch);
```
`delete1`? It's weird to have `delete2`.

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1455,6 +1455,56 @@ public void testMaxSnapshotAgeMultipleBranches() {
+    // Add B to main
+    table.newAppend().appendFile(FILE_B).commit();
+    long tAfterCommits = waitUntilAfter(table.currentSnapshot().timestampMillis());
+    Set<String> deletedFiles = Sets.newHashSet();
```
I don't think these declarations have much to do with the block of code "Add B to main" so they should be separated by an empty line.

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1455,6 +1455,56 @@ public void testMaxSnapshotAgeMultipleBranches() {
+
+    deletedFiles = Sets.newHashSet();
+    expectedDeletes = Sets.newHashSet();
+    table.expireSnapshots().expireOlderThan(testBranchHead.timestampMillis()).deleteWith(deletedFiles::add).commit();
```
Does this need a `waitUntilAfter` like you used before?

amogh-jahagirdar replied:
Ah yes, there should be a waitUntilAfter

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1455,6 +1455,56 @@ public void testMaxSnapshotAgeMultipleBranches() {
+    // Delete A on test branch
+    table.newDelete().deleteFile(FILE_A).toBranch(testBranch).commit();
+    Snapshot branchDelete = table.snapshot(testBranch);
+    table.newAppend().appendFile(FILE_C).toBranch(testBranch).commit();
```
Comment about appending C?

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1455,6 +1455,56 @@ public void testMaxSnapshotAgeMultipleBranches() {
+    //
+    expectedDeletes.add(a.manifestListLocation());
+    expectedDeletes.addAll(
+        a.allManifests(table.io()).stream().map(manifest -> manifest.path()).collect(Collectors.toSet()));
```
It looks like this `allManifests` logic may be worth putting in a private helper function.

amogh-jahagirdar replied:
Done

rdblue commented on core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java:
```diff
@@ -1455,6 +1455,56 @@ public void testMaxSnapshotAgeMultipleBranches() {
+    expectedDeletes = Sets.newHashSet();
+    table.expireSnapshots().expireOlderThan(testBranchHead.timestampMillis()).deleteWith(deletedFiles::add).commit();
+
+    //
```
???

amogh-jahagirdar replied:
Oops, probably I intended to comment something and just forgot before pushing. removed

rdblue reviewed (approved):
@amogh-jahagirdar, overall this is really close. I think that all the changes are correct and it does exactly the right thing to test branch commits by running all of the existing tests on both main and another branch. Nice work!

There are a few minor things to fix, but we should be able to get this in pretty quickly!

jackye1995 commented:
Since @rdblue also approved I will go ahead and merge this, thanks for the contribution @amogh-jahagirdar! And thanks for the review @rdblue !

amogh-jahagirdar commented:
Thanks for the review @rdblue !
